### PR TITLE
Rethink the design of the optimizer/scheduler interface.

### PR DESCRIPTION
**Pull request by haok1402**
Optimizer and scheduler construction are now imperative builders, `(cfg, ctx) -> tuple[...]` callables assigned in the training script, instead of string enums resolved opaquely inside the framework.

## Move off the string-based design

- The old `Literal["Adam", "Muon"]` enum was hard to extend. Every new behavior needed a new enum value, a resolver branch, and a config field for each knob. Anything unanticipated, like a different Muon/AdamW param split, could not be reached without editing the framework.
- Function-as-config puts customization in the training script, and arguably, in the hands of the users. To change the split, an agent forks `make_muon_optimizer` in their own script and reclassifies the params. No enum, no resolver branch. Scheduler knobs are bound with `partial(make_wsd_scheduler, ...)` next to the model.

## Follow training conventions

- Weight decay of `0.1` on both Muon and the auxiliary AdamW, per "Muon is Scalable for LLM Training" (the paper shows decaying the RMSNorm gamma matters).
- Use `AdamW`, not plain `Adam`, so weight decay is decoupled from the learning rate. This is the Megatron-LM default. See its [decoupled_weight_decay](https://github.com/NVIDIA/Megatron-LM/blob/d1410e15e164923f23b8e1a3384bf22c29640fcc/megatron/core/optimizer/optimizer_config.py#L250-L253) flag.

## Simplify the scheduler with `LambdaLR`

- WSD (warmup, stable, then cosine/linear decay) is a single `lr_lambda(step)` over `LambdaLR`, and subsumes the old cosine variant.
- Every scheduler is now a flat `LambdaLR`, so checkpoints cannot disagree structurally. The `_scheduler_structure_mismatch` check and its resume-time raise are removed.

## Correctness and Performance

This revision is purely interface-related so performance benchmark isn't necessary, but we do run again to ensure the pretraining startup logics works out properly as before.
 
```
[rank0] layer_partition: 27 layers / 2 stages -> [14, 13]
[rank0] layer_partition: 27 layers / 2 stages -> [14, 13]
2026-06-20 18:54:53 | INFO | launch(cfg=PretrainLMCfg(distributed=DistributedCfg(pipeline_parallel_size=1, context_parallel_size=1, expert_parallel_size=8, timeout=datetime.timedelta(seconds=900), sharding_strategy='fsdp'), training=TrainingCfg(dataset=PosixPath('workspace/datasets/dclm-baseline/toktxt/deepseek-v2'), sequence_length=2048, seed=1234, lr=1e-06, max_steps=25, micro_batch_size=1, global_batch_size=1024, optimizer=<function make_muon_optimizer at 0xf32e1401d20>, scheduler=<function make_constant_scheduler at 0xf32e1402140>, model=PosixPath('benchmarks/pretraining/deepseek-v2-lite/model.json'), save_interval=None, save_location=PosixPath('workspace/checkpoints/deepseek-v2-lite'), moe_load_balance_coef=0.001, moe_load_balance_type='global-batch', fp8_training='disabled', init_std=0.02, nsys_start=None, nsys_stop=None, memory_profile_start=None, memory_profile_stop=None, memory_profile_output=PosixPath('/home/haok/pith-train')), logging=LoggingCfg(wandb=None)))
2026-06-20 18:54:53 | INFO | Load checkpoint: workspace/checkpoints/deepseek-v2-lite/torch-dcp/step-00000000
2026-06-20 18:55:01 | INFO | Load checkpoint: Elapsed min=6.8s, max=7.8s
2026-06-20 18:56:14 | INFO | step 00000001/00000025 | step-time 73.231 sec | cross-entropy-loss 2.4770 | load-balance-loss 1.003646 | learning-rate 1.000000e-06 | gradient-norm 1124.9667 | tokens-per-second 28,637 | peak-gpu-memory 47.64 GB
2026-06-20 18:56:38 | INFO | step 00000002/00000025 | step-time 23.297 sec | cross-entropy-loss 2.4502 | load-balance-loss 1.004312 | learning-rate 1.000000e-06 | gradient-norm 897.0439 | tokens-per-second 90,020 | peak-gpu-memory 47.61 GB
2026-06-20 18:57:01 | INFO | step 00000003/00000025 | step-time 23.156 sec | cross-entropy-loss 2.4683 | load-balance-loss 1.004604 | learning-rate 1.000000e-06 | gradient-norm 1000.3391 | tokens-per-second 90,565 | peak-gpu-memory 47.63 GB
```

### CODE REVIEW

**Review by MasterJH5574 — APPROVED** [review]
LGTM!

**Comment by tomtang502** [comment]
Pass scheduler by make scheduler function is excellent design 👍
